File: fixtures/scale/scala/large/src/cryptobench/ciphers/simon/SimonCipher.scala
package cryptobench.ciphers.simon

import cryptobench.core.Cipher
import cryptobench.util.BlockCodec
import cryptobench.util.Bytes
import cryptobench.util.Hex

final class SimonCipher(key: SimonKey) extends Cipher {

  override def name: String = "simon"

  override def encrypt(plaintext: String): String = {
    val data = Bytes.pad(Bytes.of(plaintext), 8)
    var off = 0
    while (off < data.length) {
      BlockCodec.write(data, off, SimonRounds.encryptBlock(BlockCodec.read(data, off), key))
      off += 8
    }
    Hex.encode(data)
  }

  override def decrypt(ciphertext: String): String = {
    val data = Hex.decode(ciphertext)
    var off = 0
    while (off < data.length) {
      BlockCodec.write(data, off, SimonRounds.decryptBlock(BlockCodec.read(data, off), key))
      off += 8
    }
    Bytes.toText(data).trim
  }
}

File: fixtures/scale/dart/large/lib/ciphers/simon/simon_cipher.dart
import '../../core/cipher.dart';
import '../../util/block_codec.dart';
import '../../util/bytes.dart';
import '../../util/hex.dart';
import 'simon_key.dart';
import 'simon_rounds.dart';

class SimonCipher implements Cipher {
  final SimonKey key;

  SimonCipher(this.key);

  @override
  String name() => 'simon';

  @override
  String encrypt(String plaintext) {
    final data = padBytes(bytesOf(plaintext), 8);
    for (var off = 0; off < data.length; off += 8) {
      writeBlock(data, off, encryptBlock(readBlock(data, off), key));
    }
    return hexEncode(data);
  }

  @override
  String decrypt(String ciphertext) {
    final data = hexDecode(ciphertext);
    for (var off = 0; off < data.length; off += 8) {
      writeBlock(data, off, decryptBlock(readBlock(data, off), key));
    }
    return bytesToText(data).trim();
  }
}
